server/datastore/mysql/disk_encryption.go

```diff
@@ -0,0 +1,305 @@
+	if hasKeyCount > 0 {
+		return &fleet.BadRequestError{Message: "Key has already been escrowed for this host"}
+	}
+
+	return err
+}
+
+func (ds *Datastore) GetUnverifiedDiskEncryptionKeys(ctx context.Context) ([]fleet.HostDiskEncryptionKey, error) {
+	// NOTE(mna): currently we only verify encryption keys for macOS,
+	// Windows/bitlocker uses a different approach where orbit sends the
+	// encryption key and we encrypt it server-side with the WSTEP certificate,
+	// so it is always decryptable once received.
+	//
+	// To avoid sending Windows-related keys to verify as part of this call, we
+	// only return rows that have a non-empty encryption key (for Windows, the
+	// key is blanked if an error occurred trying to retrieve it on the host).
+	var keys []fleet.HostDiskEncryptionKey
+	err := sqlx.SelectContext(ctx, ds.reader(ctx), &keys, `
+          SELECT
+            base64_encrypted,
+            host_id,
+            updated_at
+          FROM
+            host_disk_encryption_keys
+          WHERE
+            decryptable IS NULL AND
+            base64_encrypted != ''
+	`)
+	return keys, err
+}
+
+func (ds *Datastore) SetHostsDiskEncryptionKeyStatus(
+	ctx context.Context,
+	hostIDs []uint,
+	decryptable bool,
+	threshold time.Time,
+) error {
+	if len(hostIDs) == 0 {
+		return nil
+	}
+
+	query, args, err := sqlx.In(
+		"UPDATE host_disk_encryption_keys SET decryptable = ? WHERE host_id IN (?) AND updated_at <= ?",
+		decryptable, hostIDs, threshold,
+	)
+	if err != nil {
+		return err
+	}
+	_, err = ds.writer(ctx).ExecContext(ctx, query, args...)
+	return err
+}
+
+func (ds *Datastore) GetHostDiskEncryptionKey(ctx context.Context, hostID uint) (*fleet.HostDiskEncryptionKey, error) {
+	var key fleet.HostDiskEncryptionKey
+	err := sqlx.GetContext(ctx, ds.reader(ctx), &key, `
+          SELECT
+            host_id, base64_encrypted, decryptable, updated_at, client_error
+          FROM
+            host_disk_encryption_keys
+          WHERE host_id = ?`, hostID)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			msg := fmt.Sprintf("for host %d", hostID)
+			return nil, ctxerr.Wrap(ctx, notFound("HostDiskEncryptionKey").WithMessage(msg))
+		}
+		return nil, ctxerr.Wrapf(ctx, err, "getting data from host_disk_encryption_keys for host_id %d", hostID)
+	}
+	return &key, nil
+}
+
+func (ds *Datastore) CleanupDiskEncryptionKeysOnTeamChange(ctx context.Context, hostIDs []uint, newTeamID *uint) error {
+	return ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+		return cleanupDiskEncryptionKeysOnTeamChangeDB(ctx, tx, hostIDs, newTeamID)
+	})
+}
+
+func cleanupDiskEncryptionKeysOnTeamChangeDB(ctx context.Context, tx sqlx.ExtContext, hostIDs []uint, newTeamID *uint) error {
+	// We are using Apple's encryption profile to determine if any hosts, including Windows and Linux, are encrypted.
+	// This is a safe assumption since encryption is enabled for the whole team.
```

Comment:
If you make further changes to this PR, would you mind adding a note here that we should consider basing this on the applicable config settings rather than the Apple profile (which we can leave for a future improvement).